src/ucp/core/ucp_request.c

```diff
@@ -297,6 +321,15 @@ int ucp_request_pending_add(ucp_request_t *req)
                 ucs_debug_get_symbol_name(req->send.uct.func));
 
     uct_ep = req->send.ep->uct_eps[req->send.lane];
+    if ((req->flags & UCP_REQUEST_FLAG_SEND_AM) &&
+        (req->send.msg_proto.am.header_length) &&
```

Comment:
This is common function for all type of operations, better to avoid adding branches to it.
Maybe it is possible to have a check in the every AM sending function (except zero copy)? If the operation did not succeed, then need to copy the header. Also it could be possible to keep `UCP_AM_SEND_FLAG_COPY_HEADER` set, so it can be used for header release when the operation completes

Reply:
The problem with keeping UCP_AM_SEND_FLAG_COPY_HEADER set is that you might copy the header more then once...
For example what happens when a request in the pending queue is pushed to the pending queue again? You will copy the header twice... I think this scenario is possible. 

Regarding moving the branches it's can improve performance, Performance probably more important then cleaner and safer code so maybe it's a good idea to do it. Need to do it carefully...

Reply:
Hi. Have a look again and track from where we call ucp_request_pending_add...
From what I see moving this branch to a dedicated AM function will require adding 3 new functions specifically for AM (ucp_request_am_pending_add, ucp_request_am_try_send, ucp_request_am_send) and call them accordingly.
It's will ruin the generic concept of ucp_request... functions, and all this to save one branch (checking the UCP_REQUEST_FLAG_SEND_AM flag)...

If you look on ucp_request_try_send that call ucp_request_pending_add for example, this function already use a "complex" branch...
Maybe is good to mention that calling ucp_request_pending_add means that this request is not going to be send now, and will involve SW data structure to store the request, I am not sure if this flow is part of the critical data path... 

Let me know what you think.